pkg/controllers/management/clusterupstreamrefresher/cluster_upstream_refresher.go

```diff
@@ -164,7 +163,7 @@ func (c *clusterRefreshController) refreshClusterUpstreamSpec(cluster *mgmtv3.Cl
 		apimgmtv3.ClusterConditionUpdated.False(cluster)
 		apimgmtv3.ClusterConditionUpdated.Message(cluster, fmt.Sprintf("[Syncing error%s] %s", syncFailed, err.Error()))
 
-		if upstreamConfig == nil {
+		if upstreamConfig == nil || (upstreamConfig.gkeConfig == nil && upstreamConfig.eksConfig == nil && upstreamConfig.aksConfig == nil) {
```

Comment:
Does DescribeCluster, or comparable functions for non-eks, just return a nil cluster with no error in this case?

Reply:
Would be nice to update the comment block above to account for these additional checks and the reasons we need it.

Reply:
I think I understand what you are asking and I will do my best to provide an answer:

There are 3 possibilities for what `getComparableUpstreamSpec` returns:
1. non-nil upstreamSpec and nil error -- This indicates success
2. nil upstreamSpec and non-nil error -- This indicates error and no further action should be taken (this is what caused the problem: `upstreamConfig` was not `nil`, but `upstreamConfig.eksConfig` was nil)
3. non-nil upstreamSpec and non-nil error -- This indicates that there is a problem with one of the nodegroups (like the underlying launch template has been deleted and the corresponding nodegroup needs to be recreated). Syncing is still valid in this case, but the error needs to be reported so the user can take action.

Reply:
Oh I did realize we were already checking and writing an error. So in the outage case an error is returned by the eks API?

Reply:
Yes, a 500 error is what was reported.

Reply:
No, because of 3 above. We have to report problems with node groups based on launch templates.

Reply:
Sure we can report the errors but I mean should we be updating the spec ever?

Reply:
If the AWS API reports an error then I don't think we should sync.

If Rancher detects a problem with the node groups (like the launch template was deleted) then I think we should sync. The reason we should in this case is if a launch template is deleted, then the cluster is still in a good state. The only issue is that the node group cannot be edited; it must be recreated and deleted. So, if we don't sync in this case then we could potentially cause problems when a user fixes the deleted launch template.

Consider the following scenario:
1. A user accidentally deletes a launch template on AWS.
2. The user also adds a node group from the EKS console (not from Rancher).
3. The syncer runs and sees the problem with the launch template deleted, but does not sync. The error gets reported so the user sees it in Rancher's UI.
4. The user uses Rancher to recreate the messed up node group. But Rancher doesn't see the new node group the user added from the EKS console because the sync didn't happen. So this new node group gets deleted when the user makes the change in the Rancher UI.

Reply:
Ah, and in the case the AWS API reported an error the config fields would be nil which is how can differentiate? Another alternative would be to just return nil in getComparableUpstreamSpec instead of an empty config if there's an error right? If I'm understanding this correctly then LGTM.

Reply:
Yes, we could do this error checking in the `getComparableUpstreamSpec`, but we would have to do three error checks there (one for each hosted provider) instead of one here. Also, there would still have to be checks for if "if an error is returned and the config is not nil" for each hosted provider.

In short, what I am trying to say is that even though the error checking is a little complicated here, I think it would be more so if it was moved to `getComparableUpstreamSpec`.

Reply:
Np, was just confirming that was a possible alternative to make sure I am understanding this.